### DISCUSSION TITLE
snap remove docker deletes user volumes (unexpected and harmful)

### DISCUSSION THREAD

**Discussion by christianhujer** [canonical/docker-snap#237] [General]
I had an issue with the docker snap after upgrading the kernel form 5.8 to 5.11. So I tried to upgrade the docker snap. For whatever reason I used `snap remove docker ; snap install docker` instead of `snap refresh docker`. As a consequence, I lost my volumes.

There used to be a golden rule: Uninstalling system packages does never remove user data from `/var`.

The `docker-snap` package seems to violate this golden rule.

**Comment by farshidtz** [comment]
Very late response, but it may still be relevant to some.

Unmounted container volumes are kept under Docker's root directory (`/var/snap/docker/common/var-lib-docker/volumes`). By default, SnapD makes [snapshots](https://snapcraft.io/docs/snapshots) of snap files within the [data locations](https://snapcraft.io/docs/data-locations). That means the volumes and everything else under the Docker root directory are backed up during the removal.

Here is an example on how the snapshot can be restored after the removal:
```console
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
ubuntu       latest    a04dc4851cbc   3 weeks ago   78.1MB

$ sudo snap remove docker
docker removed

$ sudo snap install docker --candidate
docker (candidate) 27.5.1 from Canonical✓ installed

$ docker images
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE

$ snap saved docker
Set   Snap    Age    Version  Rev   Size    Notes
1727  docker  48.4s  27.5.1   3064  31.2MB  auto

$ snap restore 1727
Restored snapshot #1727.

$ sudo snap restart docker
Restarted.

$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
ubuntu       latest    a04dc4851cbc   3 weeks ago   78.1MB
```
